[Ho Cheng En Bryan] iP

bryanhce opened the pull request:
# Karen Assistant
> "Your mind is for having ideas, not holding them" - David Allen ([source](https://dansilvestre.com/productivity-quotes/))

Karen Assistant frees your mind of having to remember things you need to do. It's,
- text-based
- easy to learn
- ~~FAST~~ _SUPER_ FAST to use

All you need to do is,
1. download it from [here.](https://github.com/bryanhce/ip)
2. double-click it.
3. add your tasks
4. let it manage your tasks for you 😉 

And it is **FREE**!

Features:
- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:
```Java
public class Main {
    public static void main(String[] args) {
        new Duke("src/data/tasks.txt").run();
    }
}
```

Commit:
Add Gradle support

bryanhce committed:
level 1 greet echo exit

bryanhce committed:
level 2 add list

bryanhce committed:
level 3 mark as done

bryanhce committed:
level 4 todo event deadline

bryanhce committed:
A-TextUiTesting

bryanhce committed:
level 5 handle errors

bryanhce committed:
level 6 delete

bryanhce committed:
hotfix for the throwing exceptions

bryanhce committed:
reformatting output text

bryanhce committed:
clean up code

bryanhce committed:
Reformat output string

bryanhce committed:
Add ability to save list to a file

bryanhce committed:
Save dates as date object

bryanhce committed:
Merge branch 'branch-level-7' into branch-level-8

bryanhce committed:
Reformat main code to include parser

This is to abstract some information about input text handling into
the parser class

bryanhce committed:
Add new classes TaskList and Ui

This is to allow for more OOP in the code

bryanhce committed:
Refactor code to organise them into packages

bryanhce committed:
Merge branch 'add-gradle-support'

bryanhce committed:
Add JUnit tests

bryanhce committed:
Add JavaDocs to classes and methods

bryanhce committed:
Tweak code to comply with coding standard

bryanhce committed:
Add find task feature

bryanhce committed:
Merge branch 'branch-Level-9'

bryanhce committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/duke/main/Duke.java
#	src/main/java/duke/task/Deadline.java
#	src/main/java/duke/task/Event.java

bryanhce committed:
Merge branch 'branch-A-JavaDoc'

# Conflicts:
#	src/main/java/duke/task/Deadline.java
#	src/main/java/duke/task/Task.java

bryanhce committed:
Added checkstyle with gradle

bryanhce committed:
Add duke GUI

bryanngzh commented on src/main/java/duke/main/Duke.java:
```diff
@@ -0,0 +1,125 @@
+import java.util.ArrayList;
+import java.util.Scanner;
+
+import java.time.LocalDate;
```
Great separation of import statements here! 👍

bryanngzh commented on text-ui-test/EXPECTED.TXT:
```diff
@@ -1,7 +1,15 @@
+______________________________________
+
+______________________________________
+Fine, I'll add this duke.task:
```
Seems like the duke in this line was unintended 🤔

bryanngzh commented on src/main/java/duke/main/Duke.java:
```diff
@@ -0,0 +1,125 @@
+                    String desc = parser.getTodoDescription();
+                    Todo t = new Todo(desc);
+                    tasks.addTask(t);
+                    ui.printAddTaskMessage(t, tasks.getSize());
+                    // add a deadline item
+                } else if (firstWord.equals("deadline")) {
+                    String desc = parser.getDeadlineDescription();
+                    LocalDate byDate = parser.getDeadlineDate();
+                    Deadline d = new Deadline(desc, byDate);
+                    tasks.addTask(d);
+                    ui.printAddTaskMessage(d, tasks.getSize());
+                    //add an event item
+                } else if (firstWord.equals("event")) {
+                    String desc = parser.getEventDescription();
+                    LocalDate atDate = parser.getEventDate();
+                    Event e = new Event(desc, atDate);
+                    tasks.addTask(e);
+                    ui.printAddTaskMessage(e, tasks.getSize());
+                    //delete a duke.task from the list
+                } else if (firstWord.equals("delete")) {
+                    int delValue = parser.getTaskNumber();
+                    if (delValue > tasks.getSize() || delValue <= 0) {
+                        throw new DukeException("The delete value does not exist dummy!");
+                    }
+                    Task item = tasks.getTask(delValue - 1);
+                    tasks.removeTask(delValue - 1);
+                    ui.printDeleteMessage(item, tasks.getSize());
+                    //find a task from the task list
+                } else if (firstWord.equals("find")) {
+                    String keyword = parser.getKeyword();
+                    TaskList relatedTasks = tasks.findRelatedTask(keyword);
+                    ui.printFindMessage(relatedTasks);
+                    //firstWord is not a keyword
+                } else {
+                    ui.printUnknownMessage();
+                }
+            } catch (DukeException e) {
+                ui.printErrorMessage(e.getMessage());
+            }
+        }
```
Could perhaps consider using a switch statement here? 🤔

bryanngzh commented on src/main/java/duke/task/Event.java:
```diff
@@ -0,0 +1,55 @@
+        this.atDate = atDate;
+    }
+
+    public String getEventAt() {
```
Could perhaps add JavaDocs here?

bryanngzh reviewed (commented):
Overall awesome work, love the identity/attitude you've given the bot instead of the standard Duke.
The quality of the code seems great as well.
Keep up the great work! 👍

bryanhce committed:
Merge branch 'branch-Level-10'

bryanhce committed:
Merge branch 'branch-A-CheckStyle'

changjy44 commented on src/main/java/duke/exception/DukeException.java:
```diff
@@ -0,0 +1,17 @@
+    /**
+     * Creates an instance of a Duke Exception
+     *
+     * @param str Error message
```
I think this line of code could be better with a punctuation at the end.

changjy44 commented on src/main/java/duke/main/Duke.java:
```diff
@@ -0,0 +1,117 @@
+                String keyword = parser.getKeyword();
+                TaskList relatedTasks = tasks.findRelatedTask(keyword);
+                return ui.printFindMessage(relatedTasks);
+                //firstWord is not a keyword
```
Not a coding standard comment but a switch statement might be more readable!

changjy44 commented on src/main/java/duke/main/Main.java:
```diff
@@ -0,0 +1,31 @@
+    @Override
+    public void start(Stage stage) {
+        try {
+            FXMLLoader fxmlLoader = new FXMLLoader(Main.class.getResource("/view/MainWindow.fxml"));
```
line might be long so better to leave on a new line!

changjy44 commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,131 @@
+        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("dd/MM/yyyy");
+        LocalDate atDate = LocalDate.parse(atArr[0], formatter);
+        return atDate;
+    }
```
Maybe can do javadocs for this

changjy44 commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,110 @@
+        try {
+            FileWriter fw = new FileWriter(this.filePath);
+            for (int i = 0; i < list.getSize(); i++) {
+                Task t = list.getTask(i);
```
Naming of task can be more intuitive!

changjy44 reviewed (commented):
Overall well done! I like the way you structure your code. Keep it up 😃

bryanhce committed:
Add coloured background to UI

bryanhce committed:
Add assert check

bryanhce committed:
Improve code quality of entire repository

There were methods and classes that were not inline with the coding
standard and were not high quality code.

Let's
* change the if-else statements to switch statements
* add a default branch in the switch block
* structure code logically
* make the happy path more prominent
* delete dead code

This is to ensure better code understandability.

bryanhce committed:
Merge pull request #1 from bryanhce/branch-A-Assertions

Add assert check

bryanhce committed:
Merge branch 'master' of https://github.com/bryanhce/ip

bryanhce committed:
Merge pull request #2 from bryanhce/branch-A-CodeQuality

Improve code quality of entire repository

bryanhce committed:
Merge branch 'master' of https://github.com/bryanhce/ip

bryanhce committed:
Create a yaml file for CI

bryanhce committed:
Delete github/workflows directory

bryanhce committed:
Create main.yml file

To allow for CI

bryanhce committed:
Implement an easy way to update a task

Previously, unable to update the task without deleting it and
creating a new task.

Now, task can be updated with the 'update' keyword. You can individually
update the description and date or both fields at once.

A new keyword needs to be used as this is a new operation.

bryanhce committed:
Merge branch 'branch-C-Update'

bryanhce committed:
Refactoring code by following SLAP and renaming variables for
better clarity

bryanhce committed:
Choose a personality, Karen, for Duke

bryanhce committed:
Adding user guide

bryanhce committed:
User guide format hotfix

bryanhce committed:
User guide command summary fix

bryanhce committed:
User guide command summary table fix

bryanhce committed:
Refactor code to make more readable

bryanhce committed:
Added more error handling

bryanhce committed:
Fixing nits.

Make the program close after "bye" input.
Ensure credit given to people for code.
